mono/metadata/threads.c

```diff
@@ -5065,15 +5088,67 @@ threads_add_joinable_thread_nolock (gpointer tid)
 }
 #endif
 
+static void
+threads_add_pending_joinable_runtime_thread (gpointer thread_info)
+{
+	g_assert (thread_info);
+	MonoThreadInfo *mono_thread_info = (MonoThreadInfo*)thread_info;
+
+	if (mono_thread_info->runtime_thread) {
+		gint32 old_joinable_state = mono_atomic_load_i32 (&mono_thread_info->thread_joinable_state);
+		if (old_joinable_state == THREAD_JOINABLE_STATE_NONE) {
+			if (mono_atomic_cas_i32 (&mono_thread_info->thread_joinable_state, THREAD_JOINABLE_STATE_PENDING, THREAD_JOINABLE_STATE_NONE) == THREAD_JOINABLE_STATE_NONE) {
+				gint32 result = mono_atomic_inc_i32 (&pending_joinable_runtime_thread_count);
+				g_assert (result > 0);
+			}
+		}
+	}
+}
+
+static gboolean
+threads_wait_pending_joinable_runtime_threads (uint32_t timeout)
+{
+	gboolean wait_result = TRUE;
+	if (mono_atomic_load_i32 (&pending_joinable_runtime_thread_count) > 0) {
+		joinable_threads_lock ();
+		if (timeout == MONO_INFINITE_WAIT) {
+			mono_os_cond_wait (&zero_pending_joinable_runtime_thread_event, &joinable_threads_mutex);
+		} else {
+			gint64 start = mono_msec_ticks ();
+			gint64 elapsed = 0;
+			while (mono_atomic_load_i32 (&pending_joinable_runtime_thread_count) > 0 && elapsed < timeout) {
```

Comment:
Mixing 32 and 64bits types.
Comparing signed with unsigned types.

Waiting but not checking the result of the wait.

Reply:
See above, used pretty much the implementation from mono_os_event_wait_multiple, so the issue where "inherited" from that function. Will adjust the sign/unsign, 32/64 bit in above.

Regarding not checking the result. This was intentional, the problem is that I can't know if the return value is an error or timeout, so instead I handle the spurious wakeups and timeout condition in the loop, just using the last wait result as the return of the function that will correctly indicate success or error (timeout or other error).